### PR TITLE
elixir_1_18: init at 1.18.0

### PR DESCRIPTION
**Pull request by ankhers**
## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/development/interpreters/elixir/1.18.nix [hunk 1]
```diff
@@ -0,0 +1,8 @@
+{ mkDerivation }:
+mkDerivation {
+  version = "1.18.0";
+  sha256 = "sha256-fT3J8h2uuJ+dSR58kwlUkN023yFlmTwq2/O12KbjJc4=";
+  # https://hexdocs.pm/elixir/1.17.3/compatibility-and-deprecations.html#compatibility-between-elixir-and-erlang-otp
```
Comment (Sgiath) [inline pkgs/development/interpreters/elixir/1.18.nix:5]:
```suggestion
  # https://hexdocs.pm/elixir/1.18.0/compatibility-and-deprecations.html#between-elixir-and-erlang-otp
```

Comment (ankhers) [reply pkgs/development/interpreters/elixir/1.18.nix]: Good catch, thank you!

**Comment by chiroptical** [comment]
🚀 out of curiosity, when does this become the default? I can set myself a reminder to do that if it is helpful.

**Comment by calops** [comment]
And out of even more curiosity, when does Erlang 27 become the default? We're still on 25 right now, which forces us to recompile the elixir package if we want it to use OTP27, as this combination isn't cached.

**Comment by ankhers** [comment]
I'm not sure if there was ever a full plan to describe when we should be bumping the default OTP or Elixir versions. I have been somewhat away from nixpkgs for a little while.

I think OTP 27 has been around long enough that it is worth bumping. Though, we may want to bump to 26 for a while, then to 27. I can see about putting up a PR to bump the default OTP version a little later today, or if someone else wants to do it sooner, please do and link to this ticket so everyone has an easy time finding it.

As for Elixir, it is probably worth having a discussion with the current beam maintainers in order to figure this type of thing out. I will see about a way to discuss this with as many people in the beam maintainers list as I can to see if we can come up with some kind of agreement we can follow going forward.

**Review by chiroptical — APPROVED** [review]

**Comment by calops** [comment]
My main interogation is why `pkgs.erlang` points to `pkgs.erlang25` instead of just the latest released version, as would be intuitive? The elixir package is just picking up whatever default erlang version is defined in nixpkgs.

But if we want to be conservative with these upgrades, it would maybe make sense to also distribute elixir packages compiled with the latest erlang instead of just the default one? I don't know how to name it but it sounds like a common enough use-case to me.

**Comment by mjrusso** [comment]
> I'm not sure if there was ever a full plan to describe when we should be bumping the default OTP or Elixir versions. I have been somewhat away from nixpkgs for a little while.
> 
> I think OTP 27 has been around long enough that it is worth bumping. Though, we may want to bump to 26 for a while, then to 27. I can see about putting up a PR to bump the default OTP version a little later today, or if someone else wants to do it sooner, please do and link to this ticket so everyone has an easy time finding it.
> 
> As for Elixir, it is probably worth having a discussion with the current beam maintainers in order to figure this type of thing out. I will see about a way to discuss this with as many people in the beam maintainers list as I can to see if we can come up with some kind of agreement we can follow going forward.

My apologies if this debate has already been hashed out, but is there any reason why the policy can't be to bump to the latest stable Erlang/OTP release with each Elixir package update? This release, for example, has features that use new functionality added to Erlang/OTP 27.

**Comment by ankhers** [comment]
After having some conversation in the EEF [build-and-packaging](https://the-eef.slack.com/archives/CUQVCA5K8/p1734704285632509) slack channel as well as taking conversation in the PR into consideration, I have decided to bump the default Elixir version to 1.18 as well in this PR.

I will also be making another PR later today that will bump the Erlang version to 27. I will make sure to mention this conversation in the PR in order to (hopefully) make it easier to find.

I would still like to coordinate a conversation around what our update policy should actually look like in regards to bumping the default BEAM languages versions going forward. I will be hosting this conversation on the Nix forums. If anyone that comes across this conversation would like to have a say in our policy for updating Elixir and Erlang versions, please join in and make yourself heard. I will also be posting about this on the Erlang and Elixir forums in order to get a wider audience to be able to see it. You can find that conversation [here](https://discourse.nixos.org/t/beam-language-update-policy/57682).

@Sgiath I can put up another PR later today to remove any versions of Elixir and Erlang that are currently unsupported. Thank you for pointing that out. It has unfortunately been on my todo list for quite some time.

**Comment by happysalada** [comment]
I agree with keeping the latest on both.
I do want to point to the tradeoff that packages relying on elixir or erlang might break.
When we catch breakage , it should be just a matter of notifying the maintainers and letting them do what is appropriate for their package (pinning version of elixir or patching their package)
Thank you for making this PR and starting this discussion !

**Comment by happysalada** [comment]
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 366645`

---
### `x86_64-linux`
<details>
  <summary>:x: 6 packages failed to build:</summary>
  <ul>
    <li>erlang_27</li>
    <li>gleam</li>
    <li>next-ls</li>
    <li>rabbitmq-server</li>
    <li>rabbitmq-server.doc</li>
    <li>rabbitmq-server.man</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 7 packages built:</summary>
  <ul>
    <li>elixir (elixir_1_18)</li>
    <li>elixir-ls</li>
    <li>ex_doc</li>
    <li>lexical</li>
    <li>mix2nix</li>
    <li>protoc-gen-elixir</li>
    <li>sonic-pi</li>
  </ul>
</details>

**Comment by xanderio** [comment]
> I agree with keeping the latest on both. I do want to point to the tradeoff that packages relying on elixir or erlang might break. When we catch breakage , it should be just a matter of notifying the maintainers and letting them do what is appropriate for their package (pinning version of elixir or patching their package) Thank you for making this PR and starting this discussion !

This bump of the default elixir/erlang version shouldn't then be backported to the release branches.

**Comment by happysalada** [comment]
We should also add something in the release notes.
I think its fine to fix elixir version for the packages that fail to compile

**Comment by xanderio** [comment]
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 366645`

---
### `aarch64-darwin`
<details>
  <summary>:x: 12 packages failed to build:</summary>
  <ul>
    <li>elixir (elixir_1_18)</li>
    <li>elixir-ls</li>
    <li>erlang_27</li>
    <li>ex_doc</li>
    <li>gleam</li>
    <li>lexical</li>
    <li>mix2nix</li>
    <li>next-ls</li>
    <li>protoc-gen-elixir</li>
    <li>rabbitmq-server</li>
    <li>rabbitmq-server.doc</li>
    <li>rabbitmq-server.man</li>
  </ul>
</details>

**Review by adamcstephens — CHANGES_REQUESTED** [review]
Please redo the commits in this PR. The link fix commit currently does not meet the contribution requirements for commit description, and really should be squashed with the upgrade.

I'd also like to see the changing of the default elixir version to be a discrete commit so we can backport elixir 1.18 to 24.11 without changing the default version in that release. If easier you can drop that change entirely here and we'll handle in a separate PR.

**Comment by MangoIV** [comment]
It would be great if we could change the default version to 1.18 later and first land 1.18 without requiring any breakage.

**Comment by calops** [comment]
> It would be great if we could change the default version to 1.18 later and first land 1.18 without requiring any breakage.

I agree on principle, no reason to tie the two together if we can release 1.18 right away.

> As far as I understand this is blocked by several packages that would break because of 1.18…

Would we have to fix these packages by moving them to the previous default version of elixir, or is the correct behavior to let them break and notify the maintainers to choose the best strategy (either fix upstream, patch the package or use the old elixir version)?

**Comment by MangoIV** [comment]
I doubt that the correct way is breaking rabbitmq and gleam

**Comment by happysalada** [comment]
Fixing the elixir version and notifying the maintainers is the right approach IMO

**Comment by samueltardieu** [comment]
> I doubt that the correct way is breaking rabbitmq and gleam

I've opened #367681 to make `rabbitmq-server` use Elixir 1.17.

**Comment by adamcstephens** [comment]
I've opened https://github.com/NixOS/nixpkgs/pull/367931 instead of force pushing this PR.

**Pull request closed without merging**